### PR TITLE
[SHP] Allow loading & saving Multipatch (Triangle Fan & Triangle Strip)

### PR DESCRIPTION
**Pull request by tmontaigu**

### CODE REVIEW

File: libs/qCC_io/ShpFilter.cpp [hunk 1]
```diff
@@ -144,6 +146,38 @@ static inline bool hasMeasurements(ESRI_SHAPE_TYPE shapeType)
...
 		case ESRI_SHAPE_TYPE::POLYGON_M:
 		case ESRI_SHAPE_TYPE::MULTI_POINT_M:
+		case ESRI_SHAPE_TYPE::MULTI_PATCH:
+			return true;
+		default:
+			return false;
+	}
+}
+
+enum class ESRI_PART_TYPE: int32_t {
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: int32_t is the default - I would remove this unless you have a special reason for it?

File: libs/qCC_io/ShpFilter.cpp [hunk 2]
```diff
@@ -144,6 +146,38 @@ static inline bool hasMeasurements(ESRI_SHAPE_TYPE shapeType)
...
+	TRIANGLE_FAN = 1,
+	OUTER_RING = 2,
+	INNER_RING = 3,
+	FIRST_RING = 4,
+	RING = 5,
+};
+
+static bool isValidEsriPartType(int32_t code)
+{
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Remove extra blank line.

File: libs/qCC_io/ShpFilter.cpp [hunk 3]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+			                                      scalarValues.begin() + lastIndex,
+			                                      [](ScalarType val)
+			                                      { return std::isnan(val); });
+		}
+		if (!areAllValuesOfPartsNans)
+		{
+			auto *sf = new ccScalarField("Measures");
+			if (!sf->reserveSafe(vertCount))
+			{
+				ccLog::Warning(
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Put on one line?

File: libs/qCC_io/ShpFilter.cpp [hunk 4]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+		const std::vector<int32_t> &startIndexes,
+		const std::vector<int32_t> &partTypes,
+		const std::vector<CCVector3> &points,
+		const std::vector<ScalarType> &scalarValues
+)
+{
+	size_t numParts = startIndexes.size();
+	size_t numPoints = points.size();
+
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Remove extra blank line.

File: libs/qCC_io/ShpFilter.cpp [hunk 5]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+	{
+		if (!isValidEsriPartType(partTypes[i]))
+		{
+			ccLog::Warning("[SHP] Multipatch part %d has an invalid part type (%d)", i, partTypes[i]);
+			continue;
+		}
+		auto type = static_cast<ESRI_PART_TYPE>(partTypes[i]);
+
+		const int32_t &firstIndex = startIndexes[i];
+		const int32_t &lastIndex = static_cast<const int32_t &>((i + 1 < numParts ? startIndexes[i + 1] : numPoints) -
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Again - strange line split.

File: libs/qCC_io/ShpFilter.cpp [hunk 6]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+		                                                        1);
+		const int32_t vertCount = lastIndex - firstIndex + 1;
+
+		switch (type)
+		{
+			case ESRI_PART_TYPE::TRIANGLE_STRIP:
+			{
+				ccMesh *mesh = createMesh(points, scalarValues, firstIndex, lastIndex);
+				for (int32_t j(2); j < vertCount; ++j)
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: `int32_t j(2)` -> `int32_t j = 2`

I know it's the same thing, but the latter is the form we use everywhere.

File: libs/qCC_io/ShpFilter.cpp [hunk 7]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+				{
+					mesh->addTriangle(j - 2, j - 1, j);
+				}
+				container.addChild(mesh);
+				break;
+			}
+			case ESRI_PART_TYPE::TRIANGLE_FAN:
+			{
+				ccMesh *mesh = createMesh(points, scalarValues, firstIndex, lastIndex);
+				for (int32_t j(2); j < vertCount; ++j)
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Same change -> `int32_t j = 2`

File: libs/qCC_io/ShpFilter.cpp [hunk 8]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+		return CC_FERR_NOT_ENOUGH_MEMORY;
+	}
+
+	std::vector<int32_t> partTypes = readParts(shpStream, numParts);
+	if (partTypes.empty())
+	{
+		return CC_FERR_NOT_ENOUGH_MEMORY;
+	}
+
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: More blank lines.

File: libs/qCC_io/ShpFilter.cpp [hunk 9]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+/**
+ *
+ * @param stream Output stream of the shapefile
+ * @param cloud The cloud to save (pointcloud or vertices)
+ * @param bbMing Min point of the cloud
+ * @param bbMaxg Mxx point of the cloud
+ */
+static void save3DCloud(QDataStream &stream, ccGenericPointCloud *cloud, const CCVector3d &bbMing, const CCVector3d &bbMaxg)
+{
+	unsigned numPoints = cloud->size();
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: const?

File: libs/qCC_io/ShpFilter.cpp [hunk 10]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+
+/**
+ * Tells wether the mesh's vertices are organised in triangle fan or triangle strip fashion.
+ * @param mesh
+ * @param[out] type
+ * @return CC_FERR_BAD_ENTITY_TYPE if the mesh is neither a triangle fan or triangle strip
+ */
+CC_FILE_ERROR findTriangleOrganisation(ccMesh *mesh, ESRI_PART_TYPE &type)
+{
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Blank line

File: libs/qCC_io/ShpFilter.cpp [hunk 11]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+	CCLib::VerticesIndexes *firstVert = mesh->getNextTriangleVertIndexes();
+	if (!isTriangleFan(firstVert) && !isTriangleStrip(firstVert))
+		return CC_FERR_BAD_ENTITY_TYPE;
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Blank line

File: libs/qCC_io/ShpFilter.cpp [hunk 12]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+		for (unsigned i(2); i < mesh->size(); ++i)
+		{
+			CCLib::VerticesIndexes *idx = mesh->getNextTriangleVertIndexes();
+			if (!isTriangleStrip(idx))
+				return CC_FERR_BAD_ENTITY_TYPE;
+		}
+		type = ESRI_PART_TYPE::TRIANGLE_STRIP;
+		return CC_FERR_NO_ERROR;
+
+	} else if (isTriangleFan(secondVert))
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Newline after `}`

File: libs/qCC_io/ShpFilter.cpp [hunk 13]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+	{
+		for (unsigned i(2); i < mesh->size(); ++i)
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: `unsigned i = 2`

File: libs/qCC_io/ShpFilter.cpp [hunk 14]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+		{
+			CCLib::VerticesIndexes *idx = mesh->getNextTriangleVertIndexes();
+			if (!isTriangleFan(idx))
+				return CC_FERR_BAD_ENTITY_TYPE;
+		}
+		type = ESRI_PART_TYPE::TRIANGLE_FAN;
+		return CC_FERR_NO_ERROR;
+
+	} else
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Newline after `}`

File: libs/qCC_io/ShpFilter.cpp [hunk 15]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+	stream << static_cast<int32_t>(0); // Parts
+	stream << static_cast<int32_t>(triangleType); // Parts Type
+
+	save3DCloud(stream, vertices, bbMing, bbMaxg);
+
+	qint64 recordEnd = stream.device()->pos();
+	qint64 bytesWritten = recordEnd - recordStart;
+	assert(bytesWritten == 2 * recordSize);
+	return CC_FERR_NO_ERROR;
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Blank line

File: libs/qCC_io/ShpFilter.cpp [hunk 16]
```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
...
+                                  bool load2DPolyAs3DPoly = true)
+{
+	// skip record bbox
+	shpStream.skipRawData(4 * sizeof(double));
+
+	int32_t numParts;
+	int32_t numPoints;
+	shpStream >> numParts >> numPoints;
+
+
```
Comment (asmaloney) [inline libs/qCC_io/ShpFilter.cpp]: Blank line

File: libs/qCC_io/ShpFilter.cpp (hunk 1 above)
Comment (asmaloney) [reply libs/qCC_io/ShpFilter.cpp]: Well technically `int` is the default, but...

Comment (tmontaigu) [reply libs/qCC_io/ShpFilter.cpp]: I chose int32_t mainly to be explicit as the PartType is stored as an int32_t in the file

File: libs/qCC_io/ShpFilter.cpp (hunk 5 above)
Comment (tmontaigu) [reply libs/qCC_io/ShpFilter.cpp]: Does CC source code have to follow a max line width ?
(On my IDE I set it to 120 and this line a 121 char long so it was reformated in a strange way)

File: libs/qCC_io/ShpFilter.cpp (hunk 9 above)
Comment (tmontaigu) [reply libs/qCC_io/ShpFilter.cpp]: const for the `ccGenericPointCloud` right ?

File: libs/qCC_io/ShpFilter.cpp (hunk 5 above)
Comment (asmaloney) [reply libs/qCC_io/ShpFilter.cpp]: Actually I don't know if we have one. @dgirardeau?

File: libs/qCC_io/ShpFilter.cpp (hunk 9 above)
Comment (asmaloney) [reply libs/qCC_io/ShpFilter.cpp]: I'd meant `numPoints` (I think it is const?), but if `cloud` can be too, that's good as well.

**Comment by asmaloney** [comment]
Thanks Thomas!

File: libs/qCC_io/ShpFilter.cpp (hunk 5 above)
Comment (dgirardeau) [reply libs/qCC_io/ShpFilter.cpp]: Nope, no limit for the line width ;)

**Pull request merged**